spring-ws-core/src/main/java/org/springframework/ws/client/core/observation/WebServiceObservationInterceptor.java

```diff
@@ -0,0 +1,185 @@
+        try {
+            saxParser = parserFactory.newSAXParser();
+        } catch (ParserConfigurationException e) {
+            throw new RuntimeException(e);
+        } catch (SAXException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+
+    @Override
+    public boolean handleRequest(MessageContext messageContext) throws WebServiceClientException {
+
+        TransportContext transportContext = TransportContextHolder.getTransportContext();
+        HeadersAwareSenderWebServiceConnection connection =
+                (HeadersAwareSenderWebServiceConnection) transportContext.getConnection();
+
+
+        Observation observation = WebServiceTemplateObservationDocumentation.WEB_SERVICE_TEMPLATE.start(
+                customConvention,
+                DEFAULT_CONVENTION,
+                () -> new WebServiceTemplateObservationContext(connection),
+                observationRegistry);
+
+        messageContext.setProperty(OBSERVATION_KEY, observation);
+
+        return true;
+    }
+
+    @Override
+    public void afterCompletion(MessageContext messageContext, Exception ex) throws WebServiceClientException {
+
+        Observation observation = (Observation) messageContext.getProperty(OBSERVATION_KEY);
+        Assert.notNull(observation, "Expected observation in messageContext");
+
+        WebServiceTemplateObservationContext context = (WebServiceTemplateObservationContext) observation.getContext();
+
+        WebServiceMessage request = messageContext.getRequest();
+        WebServiceMessage response = messageContext.getResponse();
+
+        if (request instanceof SoapMessage soapMessage) {
+
+            Source source = soapMessage.getSoapBody().getPayloadSource();
+            QName root = getRootElement(source);
+            if (root != null) {
+                context.setLocalPart(root.getLocalPart());
+                context.setNamespace(root.getNamespaceURI());
+            }
+            if (soapMessage.getSoapAction() != null && !soapMessage.getSoapAction().equals(TransportConstants.EMPTY_SOAP_ACTION)) {
+                context.setSoapAction(soapMessage.getSoapAction());
+            }
+        }
+
+        if (response instanceof FaultAwareWebServiceMessage faultAwareResponse) {
+            if (!faultAwareResponse.hasFault() && ex == null) {
+                context.setOutcome("success");
+            } else {
+                context.setOutcome("fault");
+            }
+        }
+
+        URI uri = getUriFromConnection();
+        if (uri != null) {
+            context.setHost(uri.getHost());
+
+            context.setContextualName("POST " + uri.getPath());
+        } else {
+            context.setContextualName("POST");
+        }
+
+        context.setError(ex);
+
+        observation.stop();
+    }
+
+    public void setCustomConvention(WebServiceTemplateConvention customConvention) {
+        this.customConvention = customConvention;
```

Comment:
I'm curious: why make a setter for this instead of take the custom convention in the constructor? It allows for users to change the custom convention at runtime, but is that something you expect users will want to do? I would err on the side of making it final and taking it in the constructor until a user requests being able to set it after construction and provides more context around why they want to do that.